File: cpp/83. Remove Duplicates from Sorted List.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
// Time: O(n), Space: O(1)
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if(!head || !head->next) return head;       
        ListNode *cur = head, *nxt = head->next;
        while(nxt) {
            if(cur->val == nxt->val) {
                cur->next = nxt->next;
                delete nxt;
                nxt = cur->next;
            } else {
                cur = cur->next;
                nxt = cur->next;
            }
        }
        return head;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}

File: python/83-Remove Duplicates from Sorted List.py
from linklist_helper import *
class Solution:
    def deleteDuplicates(self, head) -> ListNode:
        "Time: O(n), Space: O(1)"
        dummy = ListNode()
        dummy.next = head
        prev = dummy
        cur = head
        while cur:
            if prev != dummy and cur.val == prev.val:
                prev.next = cur.next
            else:
                prev = cur
            cur = cur.next
        return dummy.next

head = gen_linklist([1, 1, 2])
result = Solution().deleteDuplicates(head)
display_linklist(result)

head = gen_linklist([1, 1, 2, 3, 3])
result = Solution().deleteDuplicates(head)
display_linklist(result)

head = gen_linklist([0])
result = Solution().deleteDuplicates(head)
display_linklist(result)
